### PR TITLE
Adds ConditionalFact/Theory attributes and discoverers

### PR DESCRIPTION
**Pull request by stephentoub**
This lets us write tests like:
```C#
[ConditionalFact("RunningWithElevatedPrivileges")]
public void TestThatOnlyRunsAsAdmin() { ... }
...
private static bool RunningWithElevatedPrivileges() { ... }
```
and
```C#
[ConditionalTheory("WebSocketsAvailable")]
[InlineData("http://someUrl")]
[InlineData("https://someUrl")]
public void TestThatOnlyRunsWithWebSockets(string url) { ... }
...
private static bool WebSocketsAvailable { get { ... } }
```

cc: @davidsh, @weshaggard, @mellinoe, @Priya91, @Chrisboh 
Related to https://github.com/dotnet/corefx/issues/3349

### CODE REVIEW

File: src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        public override IEnumerable<IXunitTestCase> Discover(
+            ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute)
+        {
+            MethodInfo testMethodInfo = testMethod.Method.ToRuntimeMethod();
+
+            string conditionMemberName = factAttribute.GetConstructorArguments().FirstOrDefault() as string;
+            MethodInfo conditionMethodInfo;
+            if (conditionMemberName == null ||
+                (conditionMethodInfo = LookupConditionalMethod(testMethodInfo.DeclaringType, conditionMemberName)) == null)
```
Comment (bartonjs) [inline src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs]: You call it ConditionalMember, but only look for methods.  Why not also look for properties?

Comment (stephentoub) [inline src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs]: I look for properties: https://github.com/dotnet/buildtools/pull/336/files#diff-b4c6450d5a91549839a48eb0e0ab2185R58

Comment (bartonjs) [inline src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs]: Nevermind, you are; you've just already resolved it down to the get method.

**Comment by davidsh** [comment]
This looks nice!

Questions:

```c#
private static bool WebSocketsAvailable { get { ... } }
```
Xunit usually needs this kind of delegates to be `public`. Curious as to how you were able to use just `private`?

Will Theory w/ MemberData be supported also? i.e.

```c#
        [Theory, MemberData("EchoServers")]
        public async Task EchoBinaryMessage_Success(Uri server)
```

**Comment by davidsh** [comment]
cc: @CIPop @pgavlin

File: src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                return new[] {
+                    new ExecutionErrorTestCase(
+                        _diagnosticMessageSink,
+                        discoveryOptions.MethodDisplayOrDefault(),
+                        testMethod,
+                        "An appropriate member \"" + conditionMemberName + "\" could not be found.")
```
Comment (bartonjs) [inline src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs]: Can we explain "appropriate"?  It looks like it has to be a static 0-argument method (or get-accessor), I'm not sure if it has visibility requirements, or not.

**Comment by stephentoub** [comment]
> Xunit usually needs this kind of delegates to be public.

I'm not sure why xunit has that limitation.  In fact, I've heard recently it may have been removed in xunit bits we already have, though I haven't tried.

> Curious as to how you were able to use just private?

Private reflection is supported.

> Will Theory w/ MemberData be supported also?

Yes.  Basically my discoverer just adds its checks up front and says "I couldn't find any tests" if the condition fails, but if the condition succeeds, it just delegates all of its logic to the base TheoryDiscoverer, providing all of the features theories have.

File: src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs (hunk 2 above)
Comment (stephentoub) [inline src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs]: Sure, I can make it more descriptive.  It's basically a Boolean-returning static method or property with zero arguments, living on this or any base type.

File: src/xunit.netcore.extensions/Discoverers/ConditionalTheoryDiscoverer.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            MethodInfo conditionMethodInfo;
+            if (conditionMemberName == null ||
+                (conditionMethodInfo = ConditionalFactDiscoverer.LookupConditionalMethod(testMethodInfo.DeclaringType, conditionMemberName)) == null)
+            {
+                return new[] {
+                    new ExecutionErrorTestCase(
+                        _diagnosticMessageSink,
+                        discoveryOptions.MethodDisplayOrDefault(),
+                        testMethod,
+                        "An appropriate member \"" + conditionMemberName + "\" could not be found.")
```
Comment (bartonjs) [inline src/xunit.netcore.extensions/Discoverers/ConditionalTheoryDiscoverer.cs]: Same "appropriate" expansion comment.

**Comment by davidsh** [comment]
Would it be possible for the *Discover's to report "skipped" tests as actually skipped, in the same way as using the Xunit attribute Skip.?

This custom Xunit extension does this for example:
https://github.com/AArnott/Xunit.SkippableFact

Also, I noticed that our own ActiveIssue attribute does not report the skipped test as "Skipped" when Xunit reports all the tests pass/fail/skip.

**Comment by bartonjs** [comment]
I'd like "appropriate" to have more guidance in it (which it sounds like you're amenable to), otherwise LGTM.

**Comment by mellinoe** [comment]
LGTM

**Comment by stephentoub** [comment]
> I'd like "appropriate" to have more guidance

Done.

> Would it be possible for the *Discover's to report "skipped" tests as actually skipped, in the same way as using the Xunit attribute Skip.?

Done.

**Comment by stephentoub** [comment]
> Also, I noticed that our own ActiveIssue attribute does not report the skipped test as "Skipped" when Xunit reports all the tests pass/fail/skip.

Yes, in that case the total number of tests is just lower; essentially those discoverers just don't discover the tests, and thus they're not counted in any number.  We can have  separate conversation about whether that's the right behavior or not.  I'd be fine seeing those changed if that was the majority opinion.

**Comment by davidsh** [comment]
>Yes, in that case the total number of tests is just lower; essentially those discoverers just don't discover the tests, and thus they're not counted in any number. We can have separate conversation about whether that's the right behavior or not. I'd be fine seeing those changed if that was the majority opinion.

IMHO, I think all the variants of "skipping tests" such as Skip, ActiveIssue, ConditionalFact, etc. should report the tests as "skipped". It helps to serve as a reminder of the fact that we have those tests but are skipping them.

**Comment by davidsh** [comment]
LGTM

**Comment by weshaggard** [comment]
I agree we should count tests disabled with ActiveIssue as a skipped tests. From a consistency standpoint I would really like to have the number of tests remain constant for the runs, unless new tests are added or removed.

**Comment by bartonjs** [comment]
The problem is there's "not running this because it will fail", like [ActiveIssue] and "doesn't apply", like [PlatformSpecific].  And I'm not sure which bucket this attribute would fall into.  (Or [OuterLoop])

I'd personally like to have at a glance the number of non-passing test(case)s for my libraries; but would hate to have to figure out what the OS variant baseline was across them to have the number be meaningful.

**Pull request merged**